lib/livebook/notebook/explore/kino/pong.livemd

```diff
@@ -1,492 +1,899 @@
+      {:cont, ball} = bounce(Game.Ball.new(10, _x = 100, _y = 90, dy: 10))
+      assert ball.dy == -10
+      assert ball.y == 90
+    end
+
+    test "with ball slightly outside of scene" do
+      {:cont, ball} = bounce(Game.Ball.new(10, _x = 100, _y = 95, dy: 10))
+      assert ball.dy == -10
+      assert ball.y == 90
+    end
   end
 
-  defp clip(x, a, b), do: x |> max(a) |> min(b)
+  describe "bounces on left" do
+    test "with perfect collision on the paddle center" do
+      {:cont, ball} = bounce(Game.Ball.new(10, _x = 2 + 10, _y = 50, dx: -10))
+      assert ball.dx == 10
+      assert ball.x == 12
+    end
+
+    test "with ball slightly inside the paddle" do
+      {:cont, ball} = bounce(Game.Ball.new(10, _x = 5, _y = 50, dx: -10))
+      assert ball.dx == 10
+      assert ball.x == 12
+    end
 
-  defp handle_collisions(pong) do
-    pong
-    |> bounce_x()
-    |> bounce_y()
+    test "with the ball bottom touching the paddle top" do
+      {:cont, ball} = bounce(Game.Ball.new(10, _x = 5, _y = 31, dx: -10))
+      assert ball.dx == 10
+      assert ball.x == 12
+    end
+
+    test "with the ball top touching the paddle bottom" do
+      {:cont, ball} = bounce(Game.Ball.new(10, _x = 5, _y = 69, dx: -10))
+      assert ball.dx == 10
+      assert ball.x == 12
+    end
+
+    test "with the ball top perfectly missing the paddle bottom" do
+      assert bounce(Game.Ball.new(10, _x = 5, _y = 70, dx: -10)) == {:won, :right}
+    end
+
+    test "with the ball bottom perfectly missing the paddle top" do
+      assert bounce(Game.Ball.new(10, _x = 5, _y = 30, dx: -10)) == {:won, :right}
+    end
   end
 
-  defp bounce_y(%{ball: ball} = pong) when ball.y - @ball_r <= 0 do
-    %{pong | ball: %{ball | y: @ball_r, dy: -ball.dy}}
+  describe "bounces on right" do
+    test "with perfect collision on the paddle center" do
+      {:cont, ball} = bounce(Game.Ball.new(10, _x = 200 - 2 - 10, _y = 50, dx: 10))
+      assert ball.dx == 10
+      assert ball.x == 188
+    end
+
+    test "with ball slightly inside the paddle" do
+      {:cont, ball} = bounce(Game.Ball.new(10, _x = 195, _y = 50, dx: 10))
+      assert ball.dx == -10
+      assert ball.x == 188
+    end
+
+    test "with the ball bottom touching the paddle top" do
+      {:cont, ball} = bounce(Game.Ball.new(10, _x = 195, _y = 31, dx: 10))
+      assert ball.dx == -10
+      assert ball.x == 188
+    end
+
+    test "with the ball top touching the paddle bottom" do
+      {:cont, ball} = bounce(Game.Ball.new(10, _x = 195, _y = 69, dx: 10))
+      assert ball.dx == -10
+      assert ball.x == 188
+    end
+
+    test "with the ball top perfectly missing the paddle bottom" do
+      assert bounce(Game.Ball.new(10, _x = 195, _y = 70, dx: 10)) == {:won, :left}
+    end
+
+    test "with the ball bottom perfectly missing the paddle top" do
+      assert bounce(Game.Ball.new(10, _x = 195, _y = 30, dx: 10)) == {:won, :left}
+    end
   end
+end
+
+ExUnit.run()
```

Comment:
Sidenote: would it make sense to extend `ExUnit` API to support something like `ExUnit.run_suite(Module)`, it could automatically start too?